flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/BoundedPubSubSource.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.pubsub;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.pubsub.v1.PubsubMessage;
+
+import java.io.IOException;
+
+/**
+ * A bounded PubSub Source, similar to {@link PubSubSource} but this will stop at some point.
+ * For example after a period of being idle or and after n amount of messages have been received.
+ */
+public class BoundedPubSubSource<OUT> extends PubSubSource<OUT> {
+	private Bound<OUT> bound;
+
+	private BoundedPubSubSource() {
+		super();
+	}
+
+	protected void setBound(Bound<OUT> bound) {
+		this.bound = bound;
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		bound.start(this);
+		super.run(sourceContext);
+	}
+
+	@Override
+	void processMessage(Tuple2<PubsubMessage, AckReplyConsumer> message) {
+		super.processMessage(message);
+		bound.receivedMessage();
+	}
+
+	/**
+	 * Creates a {@link BoundedPubSubSourceBuilder}.
+	 * @param <OUT> Type of Object which will be read by the produced {@link BoundedPubSubSource}
+	 */
+	@SuppressWarnings("unchecked")
+	public static <OUT> BoundedPubSubSourceBuilder<OUT, ? extends PubSubSource, ? extends BoundedPubSubSourceBuilder> newBuilder() {
```

Comment:
This builder really only needs to be bounded by the OUT type. It seems to me that providing a more straightforward builder with less generics ceremony would increase readability and also eliminate the need for a somewhat untypical syntax with generics after the dot, e.g.  `PubSubSource.<SomeObject>newBuilder()`. Consider this:

```
public static <T> PubSubSourceBuilder<T> builderForType(Class<T> cls) {
	return new PubSubSourceBuilder<>();
}
```

```
public static final class PubSubSourceBuilder<OUT> {
  protected PubSubSource<OUT> sourceUnderConstruction;
  ...
  private String hostAndPort;

  private PubSubSourceBuilder() {
    this.sourceUnderConstruction = new PubSubSource<>();
  }

  public PubSubSourceBuilder<OUT> withHostAndPort(String hostAndPort) {
      this.hostAndPort = hostAndPort;
      return this;
    }

  public PubSubSource<OUT> build() throws IOException {
      if (credentials == null) {
        credentials = defaultCredentialsProviderBuilder().build().getCredentials();
      }
      ...
      return sourceUnderConstruction;
    }
}
```
Usage:
```
PubSubSource.builderForType(String.class).withHostAndPort("google.com/grpc:1234").build();
```

Reply:
Yes I agree @afedulov  The latest implemenation has this but I think my push and your comment just crossed each other :) Could you have a look at the current implementation again?

Reply:
:+1:  LGTM on the generics part. The question if hostAndPort should be exposed here (from the above discussion) remains open.